src/Storages/MergeTree/ReplicatedMergeMutateTaskBase.cpp

```diff
@@ -0,0 +1,225 @@
+            throw;
+        }
+
+    }
+    catch (...)
+    {
+        saved_exception = std::current_exception();
+    }
+
+
+    if (saved_exception)
+    {
+        std::lock_guard lock(storage.queue.state_mutex);
+
+        auto & log_entry = selected_entry->log_entry;
+
+        log_entry->exception = saved_exception;
+
+        if (log_entry->type == ReplicatedMergeTreeLogEntryData::MUTATE_PART)
+        {
+            /// Record the exception in the system.mutations table.
+            Int64 result_data_version = MergeTreePartInfo::fromPartName(log_entry->new_part_name, storage.queue.format_version)
+                .getDataVersion();
+            auto source_part_info = MergeTreePartInfo::fromPartName(
+                log_entry->source_parts.at(0), storage.queue.format_version);
+
+            auto in_partition = storage.queue.mutations_by_partition.find(source_part_info.partition_id);
+            if (in_partition != storage.queue.mutations_by_partition.end())
+            {
+                auto mutations_begin_it = in_partition->second.upper_bound(source_part_info.getDataVersion());
+                auto mutations_end_it = in_partition->second.upper_bound(result_data_version);
+                for (auto it = mutations_begin_it; it != mutations_end_it; ++it)
+                {
+                    ReplicatedMergeTreeQueue::MutationStatus & status = *it->second;
+                    status.latest_failed_part = log_entry->source_parts.at(0);
+                    status.latest_failed_part_info = source_part_info;
+                    status.latest_fail_time = time(nullptr);
+                    status.latest_fail_reason = getExceptionMessage(saved_exception, false);
+                }
+            }
+        }
+
+    }
+
+    return false;
+}
+
+
+bool ReplicatedMergeMutateTaskBase::executeImpl()
+{
+    MemoryTrackerThreadSwitcherPtr switcher;
+    if (merge_mutate_entry)
+        switcher = std::make_unique<MemoryTrackerThreadSwitcher>(&(*merge_mutate_entry)->memory_tracker);
+
+    switch (state)
+    {
+        case State::NEED_PREPARE :
+        {
+            prepareCommon();
+
+            /// Depending on condition there is no need to execute a merge
+            if (state == State::SUCCESS)
+                return true;
+
+            if (!prepare())
+            {
+                state = State::CANT_MERGE_NEED_FETCH;
+                return true;
+            }
+
+            state = State::NEED_EXECUTE_INNER_MERGE;
+            return true;
+        }
+        case State::NEED_EXECUTE_INNER_MERGE :
+        {
+            try
+            {
+                if (!executeInnerTask())
+                {
+                    state = State::NEED_COMMIT;
+                    return true;
+                }
+            }
+            catch (...)
+            {
+                /// Maybe part is nullptr here?
+                write_part_log(ExecutionStatus::fromCurrentException());
```

Comment:
It is definitely not cool if child class should set this callback itself anywhere

Reply:
Maybe return it as a part of prepare result?
At lease need check `if (write_part_log)`

Reply:
The whole code in all tasks requires sequential execution... For example in MutateTask all the variables are stored in a struct and various functions access them. Is not good, because there is no invariants..

Reply:
Seems like fixed